### ISSUE TITLE
Why a big jump in pulse width from 1024x to 2048x  

### ISSUE THREAD

**Issue by mokun** [mars-sim/mars-sim#527]
**Is your feature request related to a problem? Please describe.**
- While looking into how best to tweak the value `processInterval` in timePassing() in ResourceProcessing, I came across one strange phenomenon with the pulse width given by `pulse.getElapsed()`.

**Describe the situaion**
- A pulse width of 0.01153474723200986 is ideal.
- At 1024x time ratio, it can be kept at 0.0116.
- However, as soon as player increase the time ratio to x2048, the pulse width would jump up to as much as 3.37932048 and a log statement below appears 
> Proposed pulse 23.34633 clipped to a max of 3.37932048.. 

- This is because of the recent change in simulation.xml as follows: 

```xml
<!-- The time span covered by each simulation pulse is defined below. These are defined in terms of
		     milliSols. The lower the value, the more accurate the simulation but the more pulses & more expensive to calculate. Note both values can be adjusted if the processing becomes too excessive.
-->
<!-- Default: 0.0112644016 millisols ~ 1 earth seconds -->
<min-simulated-pulse value="0.0112644016"/>

<!-- Default: 3.37932048 millisols ~ 5 earth minutes or 300 earth seconds -->
<max-simulated-pulse value="3.37932048"/>
```

**Describe solution you've considered**
- Design the pulse width to increase gradually all the way up to 3.37932048 millisols, instead of making an abrupt change to jump to 3.37932048 

**Additional context**
```
00-Adir-01:180.666 (Config) MasterClock : Time-ratio 512x -> 256x
00-Adir-01:180.666 (Config) MasterClock : The scale factor becomes 28.6
00-Adir-01:185.750 (Info) ResourceProcessing : pulse.getElapsed(): 0.02306949446401972  accumulatedTime: 26.516776967189863  processInterval: 26.512727479497716
00-Adir-01:186.615 (Info) ResourceProcessing : pulse.getElapsed(): 0.02306949446401972  accumulatedTime: 26.525252192444437  processInterval: 26.512727479497716
00-Adir-01:186.661 (Info) ResourceProcessing : pulse.getElapsed(): 0.02306949446401972  accumulatedTime: 26.535783822321182  processInterval: 26.512727479497716
00-Adir-01:187.169 (Info) ResourceProcessing : pulse.getElapsed(): 0.02306949446401972  accumulatedTime: 26.521465367343964  processInterval: 26.512727479497716

00-Adir-01:418.198 (Config) MasterClock : Time-ratio 256x -> 512x
00-Adir-01:418.198 (Config) MasterClock : The scale factor becomes 32.1
00-Adir-01:428.297 (Info) ResourceProcessing : pulse.getElapsed(): 0.02306949446401972  accumulatedTime: 30.449127072798195  processInterval: 30.44871850366023
00-Adir-01:429.173 (Info) ResourceProcessing : pulse.getElapsed(): 0.02306949446401972  accumulatedTime: 30.469137045284743  processInterval: 30.44871850366023
00-Adir-01:429.196 (Info) ResourceProcessing : pulse.getElapsed(): 0.02306949446401972  accumulatedTime: 30.45659918069747  processInterval: 30.44871850366023
00-Adir-01:429.727 (Info) ResourceProcessing : pulse.getElapsed(): 0.02306949446401972  accumulatedTime: 30.465350220184284  processInterval: 30.44871850366023
00-Adir-01:429.744 (Info) ResourceProcessing : pulse.getElapsed(): 0.01730212084801479  accumulatedTime: 30.460351590200293  processInterval: 30.44871850366023

00-Adir-01:541.591 (Config) MasterClock : Time-ratio 512x -> 1024x
00-Adir-01:541.591 (Config) MasterClock : The scale factor becomes 35.7
00-Adir-01:558.709 (Info) ResourceProcessing : pulse.getElapsed(): 0.01153474723200986  accumulatedTime: 39.06610526326298  processInterval: 39.06105919775842
00-Adir-01:559.574 (Info) ResourceProcessing : pulse.getElapsed(): 0.02306949446401972  accumulatedTime: 39.07458048851751  processInterval: 39.06105919775842
00-Adir-01:559.597 (Info) ResourceProcessing : pulse.getElapsed(): 0.02306949446401972  accumulatedTime: 39.062042623930246  processInterval: 39.06105919775842
00-Adir-01:560.127 (Info) ResourceProcessing : pulse.getElapsed(): 0.01153474723200986  accumulatedTime: 39.07079366341702  processInterval: 39.06105919775842
00-Adir-01:560.150 (Info) ResourceProcessing : pulse.getElapsed(): 0.02306949446401972  accumulatedTime: 39.071562407049036  processInterval: 39.06105919775842
00-Adir-01:560.208 (Info) ResourceProcessing : pulse.getElapsed(): 0.02306949446401972  accumulatedTime: 39.06786764316929  processInterval: 39.06105919775842

00-Adir-01:824.342 (Config) MasterClock : Time-ratio 1024x -> 2048x
00-Adir-01:824.342 (Config) MasterClock : The scale factor becomes 39.3
00-Adir-01:824.388 (Config) MasterClock : [System] MasterClock : Proposed pulse 23.09256 clipped to a max of 3.37932048.
00-Adir-01:831.170 (Info) ResourceProcessing : pulse.getElapsed(): 3.37932048  accumulatedTime: 36.686148818920806  processInterval: 36.160246940652094
00-Adir-01:831.170 (Info) ResourceProcessing : pulse.getElapsed(): 3.37932048  accumulatedTime: 36.602479568049006  processInterval: 36.160246940652094
00-Adir-01:831.170 (Info) ResourceProcessing : pulse.getElapsed(): 3.37932048  accumulatedTime: 37.243603511157886  processInterval: 36.160246940652094
00-Adir-01:831.170 (Info) ResourceProcessing : pulse.getElapsed(): 3.37932048  accumulatedTime: 38.100234328304246  processInterval: 36.160246940652094

00-Adir-01:952.977 (Config) MasterClock : Time-ratio 2048x -> 1024x
00-Adir-01:952.977 (Config) MasterClock : The scale factor becomes 35.7
00-Adir-01:972.020 (Info) ResourceProcessing : pulse.getElapsed(): 0.01153474723200986  accumulatedTime: 34.30932415771873  processInterval: 34.30042900082403
00-Adir-01:972.874 (Info) ResourceProcessing : pulse.getElapsed(): 0.02306949446401972  accumulatedTime: 34.30626463574114  processInterval: 34.30042900082403
00-Adir-01:972.920 (Info) ResourceProcessing : pulse.getElapsed(): 0.02306949446401972  accumulatedTime: 34.316796265617896  processInterval: 34.30042900082403
00-Adir-01:974.096 (Info) ResourceProcessing : pulse.getElapsed(): 1.084266239808927  accumulatedTime: 34.97149315009713  processInterval: 34.30042900082403
00-Adir-01:974.096 (Info) ResourceProcessing : pulse.getElapsed(): 1.084266239808927  accumulatedTime: 34.887823899225324  processInterval: 34.30042900082403
00-Adir-01:974.096 (Info) ResourceProcessing : pulse.getElapsed(): 1.084266239808927  accumulatedTime: 34.94919239926513  processInterval: 34.30042900082403
00-Adir-01:974.316 (Info) ResourceProcessing : pulse.getElapsed(): 0.01153474723200986  accumulatedTime: 34.30826036406414  processInterval: 34.30042900082403
00-Adir-01:976.069 (Info) ResourceProcessing : pulse.getElapsed(): 0.03460424169602958  accumulatedTime: 34.33264744442952  processInterval: 34.30042900082403
00-Adir-02:006.313 (Info) ResourceProcessing : pulse.getElapsed(): 0.02306949446401972  accumulatedTime: 34.301698677660504  processInterval: 34.30042900082403
00-Adir-02:007.178 (Info) ResourceProcessing : pulse.getElapsed(): 0.01153474723200986  accumulatedTime: 34.310173902914904  processInterval: 34.30042900082403
00-Adir-02:007.213 (Info) ResourceProcessing : pulse.getElapsed(): 0.02306949446401972  accumulatedTime: 34.30917078555965  processInterval: 34.30042900082403
00-Adir-02:007.732 (Info) ResourceProcessing : pulse.getElapsed(): 0.02306949446401972  accumulatedTime: 34.30638707781432  processInterval: 34.30042900082403
00-Adir-02:007.766 (Info) ResourceProcessing : pulse.getElapsed(): 0.02306949446401972  accumulatedTime: 34.31869056867835  processInterval: 34.30042900082403

00-Adir-02:306.078 (Config) MasterClock : Time-ratio 1024x -> 2048x
00-Adir-02:306.078 (Config) MasterClock : The scale factor becomes 39.3
00-Adir-02:306.124 (Config) MasterClock [x16] : [System] MasterClock : Proposed pulse 23.34633 clipped to a max of 3.37932048.
00-Adir-02:318.754 (Info) ResourceProcessing : pulse.getElapsed(): 3.37932048  accumulatedTime: 37.181823749616534  processInterval: 37.049744103179286
00-Adir-02:318.754 (Info) ResourceProcessing : pulse.getElapsed(): 3.37932048  accumulatedTime: 38.038454566762766  processInterval: 37.049744103179286
00-Adir-02:318.754 (Info) ResourceProcessing : pulse.getElapsed(): 3.37932048  accumulatedTime: 37.14621639056526  processInterval: 37.049744103179286
```

@bevans2000 ,
Can it be done in MasterClock ?

**Comment by bevans2000** [comment]
A bit of an odd finding .

**Comment by mokun** [comment]
When clicking the speed up button to increase the time ratio, the TPS increases proportionally as well.

This shouldn't be.

When the time ratio gets above 2048x, it becomes unstable and the sleep time would go negative. 

Guess we can fix this first before fixing the pulse width.

![image](https://user-images.githubusercontent.com/1168584/145755305-b13cf5d8-484a-4df1-bfb8-a9084be8fc77.png)

**Comment by mokun** [comment]
@bevans2000 ,

We should probably fix this as well before releasing v.3.4.0. 

The simulation clock becomes unstable when going above 1024x in various scenarios.

**Comment by mokun** [comment]
One reason of becoming unstable is that, as we increase the time ratio, the TPS becomes sky rocket high (e.g. 900) and thus unsustainable. 

All we may have to do is to increase the pulse width instead and may be we limit the TPS ?

```java
	/**
	 * How many pulses per second
	 * 
	 * @return
	 */
	public double getPulsesPerSecond() {
		double ticksPerSecond = 0;

		// Make sure enough pulses have passed
		if (nextPulseId >= MAX_PULSE_LOG) {
			// Recent idx will be the previous pulse id but check it is not negative
			int recentIdx = (int)((nextPulseId-1) % MAX_PULSE_LOG);
			recentIdx = (recentIdx < 0 ? (MAX_PULSE_LOG-1) : recentIdx);

			// Oldest id will be the next pulse as it will be overwrite on next tick
			int oldestIdx = (int)(nextPulseId % MAX_PULSE_LOG);
			long elapsedMilli = (pulseLog[recentIdx] - pulseLog[oldestIdx]);
			ticksPerSecond = (MAX_PULSE_LOG * 1000D)/elapsedMilli;
		}

		return ticksPerSecond;
	}
```

**Comment by bevans2000** [comment]
A wider pulse can make the simulation less accurate as in some cases not all the pulse time is used in some Tasks. However I think the cal,ing of the ui pulses was wrong and that could be consuming too much CPU. Also I think there are some lingering ClockListeners when windows are closed.
 In #531 I made a quick change to fix the number of UI pulses they remain constants as the clock increases speed. My next task is to do a better fix where when adding a new listener an optional minimum time gap between pulses can be specified. So for ui listeners a default of 1 second could be done. For listeners involved in the simulation the minimum would be zero, I.e. no pulses are skipped.

**Comment by bevans2000** [comment]
So I can't get the rate to x1024 because the new code forces the rate down. Something must be wrong in the calculations because my laptop would work at x1024 without problems previously. 
Using parallel streams for the ClockListener we need to make sure that all the listeners have been executed before the next pulse is actioned. Previously the clock listener was synchronous but I'm not 100% sure that condition is guaranteed now. It could result in 2 concurrent calls to the Simulation clock listener which could explain the instability. 
Basically I can't reproduce the instability.

**Comment by mokun** [comment]
So if you remove calling decreaseSpeed() in a couple of places in MasterClock, can you reproduce the aforementioned instability?

**Comment by mokun** [comment]
> Using parallel streams for the ClockListener we need to make sure that all the listeners have been executed b

@bevans2000 
I just commented out calling decreaseSpeed() and the use of parallelStream() so that you can debug where the stability came from.

**Comment by bevans2000** [comment]
I still haven't had time to seriously look into this but I've had a good think.
The user wants to select a **desired rate** for the time ratio. If this is too high then the simulation decreases it but never increases if things improve. I think this is wrong. We should respect the **desired ratio** and  can not change it but report the actual achieved ratio;  obviously on a super fast machine the actual always matches the desired.
We have 2 variables that control the actual ratio:
1. Pulse Width msols passing in a pulse
2. Number of pulses per minute (or second)

The current algorithm tries to vary both variable according to the desired rate by looking at the average time taken to process the pulse. Based on that it varies the 2 above to include some idle time to be kind to the machine. One problem is the execution time of the pulses vary depending on a new sol, size of settlements, active missions and number of ui screens open.
I want to take a different and simpler approach:
- The desired rate is just that, a desired rate of the user. We never change this.
- We define an upper and lower bounds for the pulse width. The mid point is the optimal (configurable of course)
- We vary the pulses per minute to meet the desired rate. 

Ideal Pulse rate is always calculated using the optimal pulse width. But if the pulse rate is too high then the pulse width can be increased. We use the average pulse execution time to predict whether the rate is achievable.Do we care about always allowing for sleep time? Sleep time will come indirectly from the pulse rate not explicitly factored like currently. 
If the desired rate is high then the pulse rate increases. If the machine can not keep up the sleep time will suffer resulting in high CPU. This is acceptable because the user may want to advance time then slowdown. The ui reports the desired and actual.

**Comment by bevans2000** [comment]
Also using a power of 2 speed means that the TR increases quicker at higher speeds. We should drop the speed setting and just use TR instead. This will allow intermediate rates between 512, 1024 & 2048.

**Comment by bevans2000** [comment]
@mokun In this last commit 7cd6e88 (on a branch) I have stripped down the ```MasterClock``` to something much simpler. From my basic trials it seems to work beyond x1024. In this version it doesn't tried to force the TR rate but just tried the best it can. 
I'm still not sure if multiple ClockListener threads may be causing a problem but so far I can not prove this is a problem; hence I might just be being a bit paranoid.

**Comment by mokun** [comment]
> If this is too high then the simulation decreases it but never increases if things improve. I think this is wrong. We should respect the **desired ratio** and can not change it but report the actual achieved ratio;

hmmm. it actually can increase if things improve but with a caveat....

Currently, we have
1. target time ratio
2. preferred time ratio
3. actual time ratio (calculated)

The preferred time ratio becomes DIFFERENT from the target time ratio only when the player click on the plus button in an attempt to increase the time ratio.

In that case, say if the old time ratio is 1024x, then both the target time ratio and the preferred time ratio become 2048x.

However, if the system becomes unstable, then the target time will bump back to 1024x.

If a person click on the Time Tool or a button, it will attempt to try to match the target time ratio to that of the preferred one.

**Comment by mokun** [comment]
> However I don't understand what `MasterClock.getScaleFactor` is doing. It seems to bypass the ClockPulse and influences calculations in Crop & Resource Processing. 

On another thread, we were discussing how to reduce the CPU utilization and it turns out the Crop and ResourceProcess class are the major usage of CPU. 

That's where the scale factor comes in handy to reduce the CPU usage, especially in case of high time ratios.

Should these entities decide their calculates on the Pulse count or elapsed time ?

**Comment by mokun** [comment]
> We should drop the speed setting and just use TR instead. This will allow intermediate rates between 512, 1024 & 2048.

Sure. It's a good idea to have intermediate rates in-between.

That being the case, we just define the speed differently to include 768x, 1532x, etc. as well.

**Comment by bevans2000** [comment]
> > We should drop the speed setting and just use TR instead. This will allow intermediate rates between 512, 1024 & 2048.
> 
> Sure. It's a good idea to have intermediate rates in-between.
> 
> That being the case, we just have to redefine the speed differently to include 768x, 1532x, etc. as well.

Look at this 7cd6e885e5573a30c8ac734a17981c8ef15ba772 with a different implementation of the ```MasterClock```

**Comment by mokun** [comment]
> Look at this [7cd6e88](https://github.com/mars-sim/mars-sim/commit/7cd6e885e5573a30c8ac734a17981c8ef15ba772) with a different implementation of the `MasterClock`

So after you remove the scale factor, do the Crop, CompositionOfAir, ResourceProcess class increase the cpu utilization or stay almost the same as you vary the time ratio ?

**Comment by mokun** [comment]
Also, the method `getAverageTPS()` was removed. But it may have some benefit that the new method `getCurrentPulsesPerSecond()` doesn't have since the most recent TPS can be erroneous and may not be able to tell what have happened recently...

**Comment by bevans2000** [comment]
I've increased how many execution times we keep to keep a longer history. The average TPS can be derived from the execution time.

**Comment by bevans2000** [comment]
> > Look at this [7cd6e88](https://github.com/mars-sim/mars-sim/commit/7cd6e885e5573a30c8ac734a17981c8ef15ba772) with a different implementation of the `MasterClock`
> 
> So after you remove the scale factor, do the Crop, CompositionOfAir, ResourceProcess class increase the cpu utilization or stay almost the same as you vary the time ratio ?

These classes have a recalculate duration in msols. But I feel that if you want to push to a very high rate then high CPU is a consequence. I ran some tests and it seemed to behave more consistently.

**Comment by mokun** [comment]
ok. so is there anything else left to refactor in MasterClock that you can think of ?

Can you make a pull request so that I can try it out on my machine as well ?

**Comment by bevans2000** [comment]
Yes or you can do a build yourself on the branch

**Comment by mokun** [comment]
Unless you find any problems on this, can we close this now ?

**Comment by bevans2000** [comment]
Let's close it
